web-common/src/Hint/State.purs

```diff
@@ -0,0 +1,188 @@
+import Hint.Lenses (_active, _content, _mGlobalClickSubscription, _mPopperInstance, _placement)
+import Hint.Types (Action(..), Input, State, arrowRef, hintRef, popoutRef)
+import Hint.View (render)
+import Popper (OffsetOption(..), PaddingOption(..), Placement, PositioningStrategy(..), arrow, createPopper, defaultFlip, defaultModifiers, defaultPreventOverflow, destroyPopper, flipPlacement, forceUpdate, offset, pAll, preventOverflow)
+import Web.Event.Event (EventType(..))
+import Web.Event.EventTarget (addEventListener, eventListener, removeEventListener)
+import Web.HTML (HTMLElement, window)
+import Web.HTML.HTMLDocument as HTMLDocument
+import Web.HTML.HTMLElement (DOMRect, getBoundingClientRect)
+import Web.HTML.Window (document)
+import Web.UIEvent.MouseEvent as MouseEvent
+
+_hintSlot :: SProxy "hintSlot"
+_hintSlot = SProxy
+
+hint ::
+  forall slots m action.
+  MonadAff m =>
+  Array String ->
+  String ->
+  Placement ->
+  PlainHTML ->
+  ComponentHTML action ( hintSlot :: forall query. Slot query Void String | slots ) m
+hint hintElementClasses ref placement content =
+  slot
+    _hintSlot
+    ref
+    component
+    { content, placement, hintElementClasses }
+    absurd
+
+initialState :: Input -> State
+initialState { content, placement, hintElementClasses } =
+  { content
+  , placement
+  , hintElementClasses
+  , active: false
+  , mPopperInstance: Nothing
+  , mGlobalClickSubscription: Nothing
+  }
+
+component ::
+  forall m query.
+  MonadAff m =>
+  Component HTML query Input Void m
+component =
+  mkComponent
+    { initialState
+    , render
+    , eval:
+        H.mkEval
+          $ H.defaultEval
+              { initialize = Just Init
+              , finalize = Just Finalize
+              , handleAction = handleAction
+              , receive = \input -> Just $ OnNewInput input
+              }
+    }
+
+handleAction ::
+  forall m slots.
+  MonadAff m =>
+  Action ->
+  HalogenM State Action slots Void m Unit
+handleAction Init = do
+  placement <- use _placement
+  mPopperInstance <-
+    runMaybeT do
+      arrowElem <- MaybeT $ getHTMLElementRef arrowRef
+      let
+        modifiers =
+          defaultModifiers
+            <> [ arrow arrowElem (PaddingValue $ pAll 0.0)
+              , offset (OffsetValue { skidding: 0.0, distance: 8.0 })
+              , preventOverflow defaultPreventOverflow
+              , flipPlacement defaultFlip
+              ]
+      hintElem <- MaybeT $ getHTMLElementRef hintRef
+      popoutElem <- MaybeT $ getHTMLElementRef popoutRef
+      liftEffect $ createPopper hintElem popoutElem { placement, modifiers, strategy: Fixed }
+  assign _mPopperInstance mPopperInstance
+
+handleAction Finalize = do
+  mPopperInstance <- use _mPopperInstance
+  for_ mPopperInstance $ liftEffect <<< destroyPopper
+
+handleAction (OnNewInput input) = do
+  { active, content, placement } <- get
+  modify_
+    ( set _content input.content
+        <<< set _placement input.placement
+    )
```

Comment:
I've generally done this with `$` instead of brackets, and I think it looks a bit nicer. :shrug:

Reply:
will try